src/gpl/src/mbff.cpp

```diff
@@ -0,0 +1,1219 @@
+        int opt_idx = 0;
+        float opt_val = -1;
+        for (int j = 0; j < 5; j++) {
+            if (res[i][j] > opt_val) {
+                opt_val = res[i][j];
+                opt_idx = j;
+            }
+        }
+        trays[i] = start_trays[i][opt_idx];
+    }
+
+    return trays;
+}
+
+void MBFF::Remap(const std::vector<Flop> &flops,
+                 std::vector<std::vector<Tray> > &trays) {
+    int num_flops = static_cast<int>(flops.size());
+
+    std::map<int, int> small_to_large;
+    for (int i = 0; i < num_flops; i++) {
+        small_to_large[i] = flops[i].idx;
+    }
+
+    for (int i = 0; i < NUM_SIZES; i++) {
+        for (int j = 0; j < static_cast<int>(trays[i].size()); j++) {
+            for (int k = 0; k < static_cast<int>(trays[i][j].cand.size()); k++) {
+                trays[i][j].cand[k] = small_to_large[trays[i][j].cand[k]];
+            }
+        }
+    }
+}
+
+void MBFF::Run(int mx_sz, float alpha, float beta) {
+    srand(1);
+    omp_set_num_threads(num_threads);
+
+    // run k-means++ based decomposition
+    std::vector<std::vector<Flop> > pointsets = KMeansDecomp(all_flops, mx_sz);
+
+    std::vector<std::vector<Tray> > all_trays(NUM_SIZES);
+
+    for (int t = 0; t < static_cast<int>(pointsets.size()); t++) {
+
+        int num_flops = static_cast<int>(pointsets[t].size());
+
+        // run silhouette metric
+        std::vector<std::vector<Tray> > trays = RunSilh(pointsets[t]);
+
+        // run capacitated k-means per tray size
+        #pragma omp parallel for
+        for (int i = 1; i < NUM_SIZES; i++) {
+            int rows = GetRows(GetBitCnt(i)), cols = GetBitCnt(i) / rows;
+            int num_trays = (num_flops + (GetBitCnt(i) - 1)) / GetBitCnt(i);
+            for (int j = 0; j < num_trays; j++) {
+                trays[i][j].slots = GetSlots(trays[i][j].pt, rows, cols);
+            }
+            RunCapacitatedKMeans(pointsets[t], trays[i], GetBitCnt(i), 35);
+            MinCostFlow(pointsets[t], trays[i], GetBitCnt(i));
+            for (int j = 0; j < num_trays; j++) {
+                trays[i][j].slots = GetSlots(trays[i][j].pt, rows, cols);
+            }
+        }
+
+        Remap(pointsets[t], trays);
+
+        // append trays
+        for (int i = 0; i < NUM_SIZES; i++) {
+            for (int j = 0; j < static_cast<int>(trays[i].size()); j++) {
+                all_trays[i].push_back(trays[i][j]);
+            }
+        }
+    }
+
+    RunILP(all_flops, all_paths, all_trays, alpha, beta);
+}
+
+// ctor
+MBFF::MBFF(int num_flops, int num_paths, const std::vector<float> &x,
+           const std::vector<float> &y,
+           const std::vector<std::pair<int, int> > &paths, int threads,
+           utl::Logger *logger) {
+    all_flops.resize(num_flops);
+    for (int i = 0; i < num_flops; i++) {
+        odb::Point new_pt(x[i] + 70.0, y[i] + 70.0);
+
+        Flop new_flop;
+        new_flop.pt = new_pt, new_flop.idx = i;
+        all_flops[i] = new_flop;
+    }
```

Comment:
Reserve instead of resize to avoid default initialization.
```suggestion
    all_flops.reserve(num_flops);
    for (int i = 0; i < num_flops; i++) {
        odb::Point new_pt(x[i] + 70.0, y[i] + 70.0);
        all_flops.emplace_back(new_pt, i, 0.0);
    }
```